### PR TITLE
Porting the Pixel Barycentre tool in the All-in-One framework

### PR DESCRIPTION
**Pull request by amecca**
#### PR description:

This PR includes the changes necessary to run the PixelBarycentre tool from the Tracker All-in-One framework for the validation of new alignments. This allows monitoring the movements of the barycentre of BPIX (or any other structure) as part of the routine validation workflow. The outline of the proposed changes was [presented](https://indico.cern.ch/event/1455298/#47-pixelbarycentre-tool-in-the) at the Tracker Alignment meetings.

#### PR validation:

Several results were derived both with the standalone PixelBarycentre and with the new workflow, including [BPIX barycentre for 2024 pp MC](https://cms-talk.web.cern.ch/t/bpix-barycentre-for-2024-pp-mc/44713), and no discrepancies were observed.
Three unit tests were added to ensure that the tool is able to run with different configurations.

### CODE REVIEW

**Comment by perrotta** [comment]
please test

File: Alignment/OfflineValidation/plugins/PixelBaryCentreAnalyzer.cc [hunk 1]
```diff
@@ -246,322 +246,321 @@ void PixelBaryCentreAnalyzer::analyze(const edm::Event& iEvent, const edm::Event
...
+    const TrackerTopology& tkTopo = iSetup.getData(trackerTopologyToken_);
 
-    if (tkGeo->isThere(GeomDetEnumerators::PixelBarrel) && tkGeo->isThere(GeomDetEnumerators::PixelEndcap)) {
+    if (tkGeo.isThere(GeomDetEnumerators::PixelBarrel) && tkGeo.isThere(GeomDetEnumerators::PixelEndcap))
       phase_ = 0;
-    } else if (tkGeo->isThere(GeomDetEnumerators::P1PXB) && tkGeo->isThere(GeomDetEnumerators::P1PXEC)) {
+    else if (tkGeo.isThere(GeomDetEnumerators::P1PXB) && tkGeo.isThere(GeomDetEnumerators::P1PXEC))
       phase_ = 1;
-    }
+  }
```
Comment (mmusich) [inline Alignment/OfflineValidation/plugins/PixelBaryCentreAnalyzer.cc:256]: there is something funny with parenthesis going on here. I think this is superfluous now and this clause should actually be closed at L545 of the new layout. 

https://github.com/cms-sw/cmssw/blob/393e0cd2e3f3793b56b81b12c06042d1f793c651/Alignment/OfflineValidation/plugins/PixelBaryCentreAnalyzer.cc#L545

This also explains why the formatting of the following piece:

https://github.com/cms-sw/cmssw/blob/393e0cd2e3f3793b56b81b12c06042d1f793c651/Alignment/OfflineValidation/plugins/PixelBaryCentreAnalyzer.cc#L548-L564

is odd.

Comment (amecca) [reply Alignment/OfflineValidation/plugins/PixelBaryCentreAnalyzer.cc]: You're right, I decided to fix it in the first commit and rebase

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by amecca** [comment]
I see that since my modifications make `PixelBarycentreAnalyzer_cfg.py` require a config file, the test [Miscellanea](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-659605/41533/unitTests/src/Alignment/OfflineValidation/test/Miscellanea/testing.log) fails. Should I reintroduce the old hard-coded default configuration for this cfg, so that the test can run?

**Comment by mmusich** [comment]
>  Should I reintroduce the old hard-coded default configuration for this cfg, so that the test can run?

either that, or you can create a json file on-the-fly and then use it for the unit test.

**Comment by mmusich** [comment]
@cmsbuild, please test

File: Alignment/OfflineValidation/python/TkAlAllInOneTool/PixBary.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                    local["validation"]["dataset"] = local["validation"]["dataset"].format(IOV)
+                if "goodlumi" in local["validation"]:
+                    local["validation"]["goodlumi"] = local["validation"]["goodlumi"].format(IOV)
+
+                ##Write job info
+                job = {
+                    "name": "{}_{}_{}_{}_{}".format(_validationName, alignment, jobType, jobName, IOV),
+                    "dir": workDir,
+                    "exe": "cmsRun",
+                    "cms-config": "{}/src/Alignment/OfflineValidation/test/PixelBaryCentreAnalyzer_cfg.py".format(os.environ["CMSSW_BASE"]),
```
Comment (TomasKello) [inline Alignment/OfflineValidation/python/TkAlAllInOneTool/PixBary.py:53]: Would it be possible to have `PixelBaryCentreAnalyzer_cfg.py` placed in `{}/src/Alignment/OfflineValidation/python/TkAlAllInOneTool` together with the rest of python configs? Reason is simply consistency.

File: Alignment/OfflineValidation/plugins/PixelBaryCentreAnalyzer.cc
Comment (TomasKello) [inline Alignment/OfflineValidation/plugins/PixelBaryCentreAnalyzer.cc:1]: Header of this file would deserve an update. Also can we please have some README_PixBary.md with a short tutorial and documentation?

File: Alignment/OfflineValidation/test/PixelBaryCentreAnalyzer_cfg.py [hunk 3]
```diff
@@ -53,69 +111,75 @@
...
+    bcLabels_.append(label)
+    print(f'TrackerAlignment: {label=} {align=}')
+
+    for record, condition in conditions.items():
+        condition.setdefault('connect', 'frontier://FrontierProd/CMS_CONDITIONS')
+        if  (record == 'TrackerAlignmentRcd'):
+            condition.setdefault('tag', 'Alignments')
+        elif(record == 'TrackerSurfaceDeformationRcd'):
+            condition.setdefault('tag', 'Deformations')
+        elif(record == 'TrackerAlignmentErrorsExtendedRcd'): # Errors should not affect the barycentre
```
Comment (TomasKello) [inline Alignment/OfflineValidation/test/PixelBaryCentreAnalyzer_cfg.py:157]: Hm in that case do we need this?

File: Alignment/OfflineValidation/test/testingScripts/test_unitPixBary.sh [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+thistest="Alignment/PixelBarycentre single yaml configuration mp3619 (test 2/3)"
+echo "TESTING $thistest..."
+pushd test_yaml/PixBary/single/testSinglePixBary/mp3619/$runrange
+./cmsRun validation_cfg.py config=validation.json || die "Failure running $thistest" $?
+popd
+
+thistest="Alignment/PixelBarycentre single json configuration unitTest (test 3/3)"
+echo "TESTING $thistest..."
+pushd test_json/PixBary/single/testSinglePixBary/unitTest/$runrange
```
Comment (TomasKello) [inline Alignment/OfflineValidation/test/testingScripts/test_unitPixBary.sh:21]: Here I am not sure if this test would bring anything new besides testing ability of All-in-one to create job directory based on the global config in json format (while yaml format is already tested above). But I don't have a strong opinion on this, perhaps the more the better.

File: Alignment/OfflineValidation/plugins/PixelBaryCentreAnalyzer.cc
Comment (amecca) [reply Alignment/OfflineValidation/plugins/PixelBaryCentreAnalyzer.cc:1]: Done

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
@amecca do you have any plan to fix the failing unit test?

**Comment by amecca** [comment]
Yes, I will work on it next week

**Comment by TomasKello** [comment]
Hi @amecca gentle reminder. Better to close this asap.

**Comment by amecca** [comment]
Hi @mmusich and @TomasKello, sorry for waiting so long. The Jenkins logs are not available anymore, and I had problems running the unit tests offline. I remember that the failing test was `Alignment/OfflineValidation/Miscellanea`, and I fixed it. However, I am not able to check offline if all of the other tests pass.

**Comment by mandrenguyen** [comment]
+1

**Pull request merged**